File: forqs_tractlengths.py
import matplotlib.pyplot as plt
import subprocess as sp
import pickle as pkl
import pandas as pd
import numpy as np
import math

exec(open("/home/joelsmith/Projects/software/tracts/tracts.py").read())

def run_forqs(ratevec, tstart, popsize, chrlen, migscheme):
    """Runs 1 of 2 different migration scenarios.

    Scenario 1:
    A target population initially composed of individuals
    of one ancestry receives migrants from the source population at
    some rate each generation to the present

    Scenario 2:
    In the first generation, 2 populations form an admixed population with
    equal ancestry contributions (.5). Both source populations then contributions
    ancestry to this admixed population at a constant rate to the present.
    """
    codedir = "/home/joelsmith/Projects/dmis/code/"
    sp.run("rm -r -f "+codedir+"forqs_files/forqs_out",shell=True)
    if migscheme == 'source':
        for rate in ratevec:
            input_list = [''] * 27
            input_list[0] = "Trajectory_Constant id_population_size_trajectory"
            input_list[1] = "value = " + str(popsize)
            input_list[3] = "Trajectory_Constant id_migration_rate_trajectory"
            input_list[4] = "value = " + str(rate)
            input_list[6] = "PopulationConfigGenerator_LinearSteppingStone popconfig_generator"
            input_list[7] = "generation_count = " + str(tstart)
            input_list[8] = "population_count = 2"
            input_list[9] = "chromosome_pair_count = 1"
            input_list[10] = "chromosome_lengths = " + str(chrlen)
            input_list[11] = "population_size = id_population_size_trajectory"
            input_list[12] = "migration_rate:from:to = id_migration_rate_trajectory 1 2"
            input_list[14] = "RecombinationPositionGenerator_Uniform rpg"
            input_list[15] = "rate = 1"
            input_list[17] = "Reporter_Population reporter_population"
            input_list[19] = "Reporter_Timer reporter_timer"
            input_list[21] = "SimulatorConfig"
            input_list[22] = "output_directory = "+codedir+"forqs_files/forqs_out/mig_"+str(rate)+"/"
            input_list[23] = "population_config_generator = popconfig_generator"
            input_list[24] = "recombination_position_generator = rpg"
            input_list[25] = "reporter = reporter_timer"
            input_list[26] = "reporter = reporter_population"
            thefile = open(codedir+"forqs_files/forqs_in/forqs_in_mig_"+str(rate)+".txt", 'w+')
            for item in input_list:
                thefile.write("%s\n" % item)
            thefile.close()
            sp.run("rm forqs.seed",shell=True)
            sp.run("forqs "+codedir+"forqs_files/forqs_in/forqs_in_mig_"+str(rate)+".txt",shell=True)
    if migscheme == 'hybrid':
        for rate in ratevec:
            input_list = [''] * 35
            input_list[0] = "Trajectory_Constant id_population_size_trajectory"
            input_list[1] = "value = " + str(popsize)
            input_list[3] = "Trajectory_Constant id_migration_rate_0"
            input_list[4] = "value = .5"
            input_list[6] = "Trajectory_Constant id_migration_rate_1"
            input_list[7] = "value = " + str(rate)
            input_list[9] = "Trajectory_GenerationComposite id_migration_rate"
            input_list[10] = "generation:trajectory = 0 id_migration_rate_0"
            input_list[11] = "generation:trajectory = 2 id_migration_rate_1"
            input_list[13] = "PopulationConfigGenerator_LinearSteppingStone popconfig_generator"
            input_list[14] = "generation_count = " + str(tstart)
            input_list[15] = "population_count = 3"
            input_list[16] = "chromosome_pair_count = 1"
            input_list[17] = "chromosome_lengths = " + str(chrlen)
            input_list[18] = "population_size = id_population_size_trajectory"
            input_list[19] = "migration_rate:from:to = id_migration_rate 1 3"
            input_list[20] = "migration_rate:from:to = id_migration_rate 2 3"
            input_list[22] = "RecombinationPositionGenerator_Uniform rpg"
            input_list[23] = "rate = 1"
            input_list[25] = "Reporter_Population reporter_population"
            input_list[27] = "Reporter_Timer reporter_timer"
            input_list[29] = "SimulatorConfig"
            input_list[30] = "output_directory = "+codedir+"forqs_files/forqs_out/mig_"+str(rate)+"/"
            input_list[31] = "population_config_generator = popconfig_generator"
            input_list[32] = "recombination_position_generator = rpg"
            input_list[33] = "reporter = reporter_timer"
            input_list[34] = "reporter = reporter_population"
            thefile = open(codedir+"forqs_files/forqs_in/forqs_in_mig_"+str(rate)+".txt", 'w+')
            for item in input_list:
                thefile.write("%s\n" % item)
            thefile.close()
            sp.run("rm forqs.seed",shell=True)
            sp.run("forqs "+codedir+"forqs_files/forqs_in/forqs_in_mig_"+str(rate)+".txt",shell=True)

def get_forqs_tracts(ratevec, chrlen, maxlen, npts, migscheme):
    bins = np.array([np.arange(0,maxlen*(1+.5/npts),float(maxlen)/npts)])[0]
    bintable = pd.DataFrame(index=bins[:-1],columns=ratevec)
    for rate in ratevec:
        data = get_data(rate, migscheme)
        chromdict = make_chroms(data['chunks'],chrlen,rate)
        tractdict = make_tracts(chromdict,chrlen)
        tractbins = make_tractbins(tractdict,chrlen,maxlen,npts)
        popahist = tractbins.get('popahist')
        bintable[rate] = popahist/np.sum(popahist)
    bindir = '/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp/'
    pkl.dump(bintable,open(bindir+"forqs_bintable.pkl", "wb" ))

def get_data(rate, migscheme):
    tempdir = "/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp/"
    outdir = "/home/joelsmith/Projects/dmis/code/forqs_files/forqs_out/mig_"+str(rate)+"/"
    if migscheme == 'hybrid':
        sp.run("tail -n+3 "+outdir+"population_final_pop3.txt > "+tempdir+"pop_chunks.txt",shell=True)
    if migscheme == 'source':
        sp.run("tail -n+3 "+outdir+"population_final_pop2.txt > "+tempdir+"pop_chunks.txt",shell=True)
    sp.run("sed 's/[^0-9]/ /g' "+tempdir+"pop_chunks.txt > "+tempdir+"pop_data.txt",shell=True)
    sp.run("cat "+tempdir+"pop_data.txt | awk '{ print NF }' > "+tempdir+"pop_nums.txt",shell=True)
    longrow = max(open(tempdir+"pop_data.txt", 'r'), key=len)
    cn = list(range(1,len(longrow.split())+1))
    chunks = pd.read_table(tempdir+"pop_data.txt", names=cn, delim_whitespace=True)
    chunks = chunks.as_matrix()
    poptext = np.loadtxt(tempdir+"pop_nums.txt")
    popnums = poptext[np.nonzero(poptext)]
    sp.run("rm "+tempdir+"pop_nums.txt",shell=True)
    sp.run("rm "+tempdir+"pop_data.txt",shell=True)
    sp.run("rm "+tempdir+"pop_chunks.txt",shell=True)
    data = {'chunks':chunks,'popnums':popnums}
    return(data)

def make_chroms(chunks, chrlen, rate):
    indpos = chunks[:,2::2]
    pos =  np.int_([value for value in np.unique(indpos) if not math.isnan(value)])
    ids = chunks[:,1::2]
    chrommatrix = np.zeros(shape=(len(ids),len(pos)))
    for row in range(chrommatrix.shape[0]):
        chunkindex = len(indpos[row,np.logical_not(np.isnan(indpos[row,]))])
        if chunkindex==0:
            chrommatrix[row,:] = ids[row,0]
        chunkstart = 0
        for chunk in range(chunkindex):
            chunkend = list(np.where(pos==indpos[row,chunk])[0])[0]
            chrommatrix[row,chunkstart:chunkend] = ids[row,chunk]
            chunkstart = chunkend
        truevalues = ids[row,np.logical_not(np.isnan(ids[row,]))]
        chrommatrix[row,np.where(chrommatrix[row,]==0)] = truevalues[-1]
    chroms = np.copy(chrommatrix)
    n = chrommatrix.shape[0]
    chroms[np.where(chroms<(n-1))] = 1
    chroms[np.where(chroms>(n-1))] = 2
    haptable = pd.DataFrame(data = chroms,columns=pos,dtype='i')
    tempdir = "/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp/"
    hapfile = tempdir+"haptable_mig_"+str(rate)+".csv"
    haptable.to_csv(hapfile,index=False)
    pos = np.append(pos,chrlen)
    chromdict = {"chroms":chroms,"pos":pos}
    return(chromdict)

def make_tracts(chromdict,chrlen):
    chroms = chromdict['chroms']
    pos = chromdict['pos']
    switches = np.apply_along_axis(np.diff,1,chroms)
    switchvec = np.apply_over_axes(np.count_nonzero,switches,1)
    tractstot = np.sum(switchvec)+len(np.nonzero(switchvec)[0])
    switchmatrix = switches[np.nonzero(switchvec)[0],]
    hapsmatrix = chroms[np.nonzero(switchvec)[0],]
    tractsmatrix = np.zeros((tractstot,2),dtype='int')
    noswitches = chroms[np.where(switchvec==0),0][0]
    fulltracts = np.transpose(np.vstack((np.repeat(chrlen,len(noswitches)),noswitches)))
    index = 0
    for row in range(switchmatrix.shape[0]):
        switchsites = np.logical_not(switchmatrix[row,]==0)
        switchpos = pos[np.append(np.where(switchsites),len(pos)-1)]
        switchpos = np.append(0,switchpos)
        hapids = hapsmatrix[row,np.append(np.where(switchsites),len(pos)-2)]
        haplengths = np.diff(switchpos)
        hapstack = np.transpose(np.vstack((haplengths,hapids)))
        tractsmatrix[index:(index+len(hapids)),] = hapstack
        index = index + (len(hapids))
    tractsmatrix = np.vstack((tractsmatrix,fulltracts))
    tractsmatrix = tractsmatrix.astype(int)
    popa = tractsmatrix[np.where(tractsmatrix[:,1]==1),]
    popb = tractsmatrix[np.where(tractsmatrix[:,1]==2),]
    tractdict = {"popa":popa,"popb":popb}
    return(tractdict)

def make_tractbins(tractdict,chrlen,maxlen,npts):
    bins = np.array([np.arange(0,maxlen*(1+.5/npts),float(maxlen)/npts)])[0]
    popa = np.reshape(tractdict['popa'],(tractdict['popa'].shape[1],2))
    popb = np.reshape(tractdict['popb'],(tractdict['popb'].shape[1],2))
    popaMorgans = popa[:,0]/chrlen
    popbMorgans = popb[:,0]/chrlen
    popahist = np.histogram(popaMorgans,bins,density=True)[0]
    popbhist = np.histogram(popbMorgans,bins,density=True)[0]
    tractbins = {'popahist':popahist,'popbhist':popbhist}
    return(tractbins)

def plot_figure_forqs(ratevec,tstart,Ls,pop,maxlen,npts,migscheme):
    """Plots the predicted tract length distributions against the
    simulated tract length distributions from forqs."""
    bindir = '/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp/'
    bintable = pkl.load(open(bindir+"forqs_bintable.pkl", "rb"))
    bins = np.append(bintable.index.values,1)
    rows = bintable.shape[0]
    xAxis = np.array(list(range(1,rows+1)))[:rows-1] / np.array(rows)
    for rate in ratevec:
        if migscheme=='hybrid':
            mig = continuous_mig_hybridzone(rate,tstart)
        if migscheme=='source':
            mig = continuous_mig(rate,tstart)
        model = demographic_model(mig)
        tractsOut = model.expectperbin(Ls,pop,bins)
        tractDist = np.array(tractsOut)
        relTractDist = tractDist / np.sum(tractDist)
        obs = np.asarray(np.where(np.logical_not(bintable[rate]==0)))[0][:-1]
        obsvalues = bintable.iloc[obs][rate]
        plt.plot(bins[:npts],np.log(relTractDist[:npts:]),label='_nolegend_')
        plt.scatter(x=bintable.index[obs],y=np.log(obsvalues))
        plt.xlabel('Tract Length (Morgans)')
        plt.ylabel('log(Relative Frequency)')
        plt.legend(loc='upper right',title='Migration Rate')

File: forqs_tractlengths.r

run.multiple.simulations = function(params, mig.vec) {
    temploc = '/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp/'
    for (i in 1:length(mig.vec)) {
        params$migration.rate = mig.vec[i]   
        run.forqs.simulation(params)
        chroms.matrix = get.chroms(params)
        tract.table.list = get.forqs.tract.length.distr(chroms.matrix)
        save(tract.table.list,file=paste(temploc,'tracts_',mig.vec[i],'.RDATA',sep=''))
        rm(chroms.matrix)
        rm(tract.table.list)
    }
}

plot.multiple.simulations = function(params, mig.vec) {
    forqs.list = list()
    temploc = '/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp/'
    for (i in 1:length(mig.vec)) {
        load(paste(temploc,'tracts_',mig.vec[i],".RDATA",sep=''))
        tract.bin.list = get.tract.length.bins(tract.table.list,params)
        tract.bin.plot = tract.bin.list[[1]][,2]/sum(tract.bin.list[[1]][,2])
        forqs.list[[i]] = tract.bin.plot
    }
    return(forqs.list)
}
   
run.forqs.simulation = function(params) {
    options(scipen=10)    
    outdir = paste(params$output.file,"mig_",params$migration.rate,sep='')
    forqs.input = vector()
    forqs.input[1]   = "Trajectory_Constant id_population_size_trajectory"
    forqs.input[2]  = paste("value = ",params$population.size,sep='')
    forqs.input[4]   = "Trajectory_Constant id_migration_rate_trajectory"
    forqs.input[5]  = paste("value = ",params$migration.rate,sep='')
    forqs.input[7]   = "PopulationConfigGenerator_LinearSteppingStone popconfig_generator"
    forqs.input[8]  = paste("generation_count = ",params$generations,sep='')   
    forqs.input[9]  = "population_count = 2"
    forqs.input[10] = "chromosome_pair_count = 1"
    forqs.input[11] = paste("chromosome_lengths = ",params$chromosome.length,sep='')   
    forqs.input[12] = "population_size = id_population_size_trajectory"   
    forqs.input[13] = "migration_rate:from:to = id_migration_rate_trajectory 1 2"          
    forqs.input[15] = "RecombinationPositionGenerator_SingleCrossover rpg"
    forqs.input[17] = "Reporter_Population reporter_population"
    forqs.input[19] = "Reporter_Timer reporter_timer"
    forqs.input[21] = "SimulatorConfig"  
    forqs.input[22] = paste("output_directory = ",outdir,sep='')    
    forqs.input[23] = "population_config_generator = popconfig_generator"              
    forqs.input[24] = "recombination_position_generator = rpg"    
    forqs.input[25] = "reporter = reporter_timer"    
    forqs.input[26] = "reporter = reporter_population"     
    forqs.input[which(is.na(forqs.input))] = " "
    input.file = "/home/joelsmith/Projects/dmis/code/forqs_files/forqs_in/forqs_source_target.txt"
    write(forqs.input,file=input.file,ncolumns=1)      
    system("rm forqs.seed") 
    system(paste("rm -r -f ",params$output.file,sep=''))
    system(paste("forqs ",input.file,sep=''))
}

get.forqs.tract.length.distr = function(chroms.matrix) {
    pos = chroms.matrix[1,]
    haps.matrix = chroms.matrix[-1,]    
    tract.table = c(0,0)
    for (i in 1:nrow(haps.matrix)) {
        switches = which(diff(haps.matrix[i,])!=0)
        switches.sites = c(1,switches,ncol(haps.matrix))
        hap.ids = haps.matrix[i,switches.sites[-1]]
        switches.inner.pos = pos[switches] + round((pos[switches+1] - pos[switches])/2) 
        hap.lengths = diff(c(0,switches.inner.pos,pos[length(pos)]))
        tract.table = rbind(tract.table,cbind(hap.lengths,hap.ids))
    }
    tract.table = tract.table[-1,]
    tracts.a = tract.table[which(tract.table[,2]==1),]
    tracts.b = tract.table[which(tract.table[,2]==2),]
    tract.table.list = list("tracts.a" = tracts.a,"tracts.b" = tracts.b)
    for (i in 1:2) {
        tract.table = tract.table.list[[i]]
        tract.bins = unique(tract.table[,1])
        tract.length.table = cbind(tract.bins,rep(0,length(tract.bins)))
        for (j in 1:length(tract.bins)) {
            tract.length.table[j,2] = length(which(tract.table[,1]%in%tract.bins[j]==TRUE))
        }
        tract.table.list[[i]] = tract.length.table[order(tract.length.table[,1]),]
   }
   return(tract.table.list)
}
   
get.chroms = function(params) {
    outdir = paste(params$output.file,"mig_",params$migration.rate,sep='')
    temp.loc = "/home/joelsmith/Projects/dmis/code/forqs_files/forqs_temp"
    system(paste("tail -n+3 ",outdir,"/population_final_pop2.txt > ",temp.loc,"/pop_chunks.txt",sep=''))            
    system(paste("sed 's/[^0-9]/ /g' ",temp.loc,"/pop_chunks.txt > ",temp.loc,"/pop_data.txt",sep=''))
    system(paste("cat ",temp.loc,"/pop_data.txt | awk '{ print NF }' > ",temp.loc,"/pop_nums.txt",sep=''))        
    cn <- paste("V",1:max(scan(paste(temp.loc,"/pop_nums.txt",sep=''))),sep='')
    chunks = read.table(file = paste(temp.loc,"/pop_data.txt",sep=''), col.names = cn,fill=TRUE)
    chunks = no.list.func(chunks)
    chroms.list = make.chroms(chunks)
    chroms.matrix = chroms.list[[1]]  
    pos = c(chroms.list[[2]],params$chromosome.length)
    n = nrow(chroms.matrix)
    chroms.matrix[is.element(chroms.matrix,c(0:(n-1)))] = 1
    chroms.matrix[is.element(chroms.matrix,c(n:(n*2)))] = 2
    chroms.matrix = rbind(pos,chroms.matrix)
    return(chroms.matrix)    
}
         
plot.chroms = function(chroms.matrix) {
    colors = c("slateblue","sienna1")
    image(t(chroms.matrix[-1,]),col=colors) 
}    
     
get.tract.length.bins = function(tract.table.list,params) {
    options(scipen=10)      
    for (j in 1:2) {
        tract.length.distr = tract.table.list[[j]]
        bins = seq(1,params$chromosome.length,params$bin.size)
        bin.table = cbind(bins,c((bins[-1]-1),params$chromosome.length))
        new.bins = matrix(nrow=nrow(bin.table),ncol=2)
        new.bins[,1] = bin.table[,2]#-(params$bin.size/2)
        for (i in 1:nrow(bin.table)) { 
            uppers = which(tract.length.distr[,1] >= bin.table[i,1])
            lowers = which(tract.length.distr[,1] <= bin.table[i,2])
            new.bins[i,2] = sum(tract.length.distr[intersect(uppers,lowers),2])      
        }  
        tract.table.list[[j]] = new.bins
    }
    return(tract.table.list)
}     
                                                               
make.chroms = function(data) {
    ind.pos = data[,seq(3,ncol(data),by=2)]
    pos = c(0,unique(sort(as.numeric(ind.pos[which(ind.pos != "NA")]))))
    ids = data[,seq(2,ncol(data),by=2)]
    chrom.matrix = matrix(nrow = nrow(data),ncol = (length(pos)+1))
    for (i in 1:nrow(data)) {
        chunk.index = length(which(ind.pos[i,]!="NA"))
        if (chunk.index==0) {
            chrom.matrix[i,] = as.numeric(ids[i,1])
            next
        }
        chunk.start = 1
        for (j in 1:chunk.index) {
            chunk.end = which(pos == ind.pos[i,j])
            chrom.matrix[i,chunk.start:chunk.end] = as.numeric(ids[i,j])
            chunk.start = chunk.end   
        }
        chrom.matrix[i,c(max(which(!is.na(chrom.matrix[i,]))):ncol(chrom.matrix))] = ids[i,max(which(ids[i,]!="NA"))]
    }
    chrom.matrix = apply(chrom.matrix,2,as.numeric)
    chrom.list = list("chrom.matrix"=chrom.matrix,"pos"=pos)
    return(chrom.list)
}

no.list.func = function(data) {
    new.matrix = matrix(nrow = nrow(data), ncol = ncol(data))
	for (i in 1:ncol(data)) {
        new.matrix[,i] = (paste(unlist(data[,i])))
    }
	return(new.matrix)
}

#---------------------Forqs plots--------------------------------------------
source("forqs_tractlengths.r")

params = list(  
    "population.size" = 1000,
    "generations" = 30,
    "chromosome.length" = 100000000,
    "output.file" = "/home/joelsmith/Projects/dmis/code/forqs_files/forqs_out/",
    "bin.size" = 500000)

mig.vec = c(.001,.03,.05)

run.multiple.simulations(params,mig.vec)
binlist = plot.multiple.simulations(params,mig.vec)
binmatrix = matrix(unlist(binlist),ncol=3,byrow=TRUE)
temploc = '/Projects/dmis/code/forqs_files/forqs_temp/'
write.csv(binmatrix,file=paste(temploc,'binmatrix.csv',sep=''),col.names=FALSE)    




